OMCompiler/SimulationRuntime/c/simulation/solver/gbode_main.c

```diff
@@ -0,0 +1,1973 @@
+        infoStreamPrint(LOG_SOLVER, 0, "gbodef_main: Failed to calculate step at time = %5g.", gbfData->time);
+        if (gbfData->stepSize > MINIMAL_STEP_SIZE) {
+          // Try smaller steps, if possible.
+          gbfData->stepSize = gbfData->stepSize / 2.;
+          infoStreamPrint(LOG_SOLVER, 0, "Try half of the step size = %g", gbfData->stepSize);
+          err = 100;
+          continue;
+        } else {
+          errorStreamPrint(LOG_STDOUT, 0, "Simulation abborted!");
+          messageClose(LOG_SOLVER);
+          return -1;
+        }
+      }
+
+      for (i = 0, err=0; i < nFastStates; i++) {
+        ii = gbData->fastStates[i];
+        // calculate corresponding values for the error estimator and step size control
+        gbfData->errtol[ii] = Rtol * fmax(fabs(gbfData->y[ii]), fabs(gbfData->yt[ii])) + Atol;
+        gbfData->errest[ii] = fabs(gbfData->y[ii] - gbfData->yt[ii]);
+        gbfData->err[ii] = gbfData->tableau->fac * gbfData->errest[ii] / gbfData->errtol[ii];
+        err = fmax(err, gbfData->err[ii]);
+      }
+
+      gbData->err_fast = err;
+      gbfData->errValues[0] = err;
+      gbfData->stepSizeValues[0] = gbfData->stepSize;
+
+      // Store performed stepSize for adjusting the time in case of latter interpolation
+      // Call the step size control
+      gbfData->lastStepSize = gbfData->stepSize;
+      gbfData->stepSize *= gbfData->stepSize_control(gbfData->errValues, gbfData->stepSizeValues, gbfData->tableau->error_order);
+
+      // debug ring buffer for the states and derviatives of the states
+      if (ACTIVE_STREAM(LOG_GBODE_V)) {
+        infoStreamPrint(LOG_GBODE_V, 1, "ring buffer during steps of inner integration");
+        infoStreamPrint(LOG_GBODE_V, 0, "old value:");
+        printVector_gb(LOG_GBODE_V, "y", gbfData->yOld, nStates, gbfData->time);
+        debugRingBuffer(LOG_GBODE_V, gbfData->x, gbfData->k, nStates, gbfData->tableau, gbfData->time, gbfData->lastStepSize);
+        infoStreamPrint(LOG_GBODE_V, 0, "new value:");
+        printVector_gb(LOG_GBODE_V, "y", gbfData->y, nStates, gbfData->time + gbfData->lastStepSize);
+        messageClose(LOG_GBODE_V);
+      }
+
+      // Re-do step, if error is larger than requested
+      if (err > 1)
+      {
+        gbfData->stats.nErrorTestFailures++;
+        infoStreamPrint(LOG_SOLVER, 0, "Reject step from %10g to %10g, error %10g, new stepsize %10g",
+                        gbfData->time, gbfData->time + gbfData->lastStepSize, err, gbfData->stepSize);
+        if (ACTIVE_STREAM(LOG_GBODE_STATES)) {
+          dumpFastStates_gbf(gbData, gbfData->time + gbfData->lastStepSize);
+        }
+      }
+    } while (err > 1);
+
+    // Count succesful integration steps
+    gbfData->stats.nStepsTaken += 1;
+
+    // interpolate the slow states to the boundaries of current integration interval, this is used for event detection
+    // interpolate the slow states on the time of the current stage
+    if (gbfData->interpolation==1) {
```

Comment:
Make `gbfData->interpolation` an enum or make it super clear that this means the order of the interpolation method. An enum is better if we want to add multiple interpolation methods with the same order.